### DISCUSSION TITLE
Managing storage space for packages

### DISCUSSION THREAD

**Discussion by jfolz** [devpi/devpi#886] [Q&A]
Hi,

When operating a caching PyPI mirror, how does one go about managing storage for the cached packages?
Reason I'm asking is that we're working mostly using data science packages and those can take up a lot of space quickly. Is there a way to LRU delete files?

**Comment by fschulze** [comment]
``devpi remove`` works for mirrored files (or ``DELETE`` with http). The problem is how to determine least recently used. That is why cleanup isn't built in. One way would be to analyze log files for that.

**Comment by Arpafaucon** [comment]
chiming in because I've spent some time reading the doc looking for this exact answer. I'm setting up `devpi` for our organization and I'm concerned about the fact that `root/pypi` storage never goes down. Am I right to predict that if (as hinted by OP) some person randomly downloads a very heavy and rarely used wheel, that's 1Go sitting on the disk forever ?

If that helps move forward the conversation, I guess that for our use case, a simple `devpi purge-mirror` command that removes **every** wheel we have for `root/pypi` would already be great. We'd run that once a month to purge it, and the very first job after will have to warm the cache again. If my understanding is correct, we'd need #882 / https://github.com/devpi/devpi/discussions/937 as a first step to implement it.

for context:
- Our use case in to use `devpi` to reduce load on `pypi` and improve our build times (notably in CI, where caching is hard to setup at some places). So, 100% the "caching proxy" part, although we're considering migrating our internal packages from `nexus` once we get more familiar with devpi.
- we download internally ~300 packages, for a total of ~100k requests / day

**Reply by fschulze** [reply]
Adding that as an option to the existing json views in devpi-server should be relatively straight forward. You basically need to look at what is done when upstream can't be reached and stale data is served. For this use case options in the query string would suffice (like ``no_projects`` currently), no need for POSTing json. There should also be no performance issue, as all the data should already exist in the database. The mentioned tickets would require further work in devpi-client/web, but for your use case the json data should be enough. A basic implementation for ``purge-mirror`` could be done with a script which would call ``devpi remove`` or the HTTP equivalent for every cached project.

**Reply by fschulze** [reply]
There is an ``offline`` (or similar) flag on the ``MirrorIndex`` class which causes returning stale data.

**Reply by Arpafaucon** [reply]
ok thanks for the pointers ! I'll be off-work for the next month, and I'll see how the server fares when I return. If relevant I'll try my luck at implementing something
